pkg/clusteragent/autoscaling/workload/local/replica_calculator.go

```diff
@@ -0,0 +1,325 @@
+	recommendationTimestamp time.Time
+}
+
+func newReplicaCalculator(podWatcher common.PodWatcher) replicaCalculator {
+	return replicaCalculator{
+		PodWatcher: podWatcher,
+	}
+}
+
+// CalculateHorizontalRecommendations is the entrypoint to calculate the horizontal recommendation for a given DatadogPodAutoscaler
+func (r replicaCalculator) CalculateHorizontalRecommendations(dpai model.PodAutoscalerInternal, lStore loadstore.Store) (*model.ScalingValues, error) {
+	currentTime := time.Now()
+
+	// Get current pods for the target
+	targetRef := dpai.Spec().TargetRef
+	targets := dpai.Spec().Targets
+	targetGVK, targetErr := dpai.TargetGVK()
+	if targetErr != nil {
+		return nil, fmt.Errorf("Failed to get GVK for target: %s, %s", dpai.ID(), targetErr)
+	}
+
+	podOwnerName := targetRef.Name
+	namespace := dpai.Namespace()
+
+	podOwner := common.NamespacedPodOwner{
+		Namespace: namespace,
+		Name:      podOwnerName,
+		Kind:      targetGVK.Kind,
+	}
+	pods := r.PodWatcher.GetPodsForOwner(podOwner)
+	if len(pods) == 0 {
+		// If we found nothing, we'll wait just until the next sync
+		return nil, fmt.Errorf("No pods found for autoscaler: %s, gvk: %s, name: %s", dpai.ID(), targetGVK.String(), targetRef.Name)
+	}
+
+	recommendedReplicas := model.HorizontalScalingValues{}
+
+	for _, target := range targets {
+		recSettings, err := newResourceRecommenderSettings(target)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to get resource recommender settings: %s", err)
+		}
+
+		queryResult := lStore.GetMetricsRaw(recSettings.metricName, namespace, podOwnerName, recSettings.containerName)
+		rec, utilizationRes, err := recSettings.recommend(currentTime, pods, queryResult, float64(*dpai.CurrentReplicas()))
+		if err != nil {
+			log.Debugf("Got error calculating recommendation: %s", err)
+			break
+		}
+		ts := utilizationRes.recommendationTimestamp
+
+		telemetryHorizontalLocalUtilizationPct.Set(
+			float64(utilizationRes.averageUtilization),
+			namespace,
+			podOwnerName,
+			dpai.Name(),
+			string(recommendedReplicas.Source),
+			le.JoinLeaderValue,
+		)
+
+		// Always choose the highest recommendation given
+		if rec > recommendedReplicas.Replicas {
+			recommendedReplicas.Replicas = rec
+			recommendedReplicas.Timestamp = ts
+			recommendedReplicas.Source = datadoghq.DatadogPodAutoscalerLocalValueSource
+		}
+	}
+
+	if recommendedReplicas.Replicas == 0 {
+		return nil, fmt.Errorf("No recommendation found for autoscaler: %s", dpai.ID())
+	}
+
+	telemetryHorizontalLocalRecommendations.Set(
+		float64(recommendedReplicas.Replicas),
+		namespace,
+		podOwnerName,
+		dpai.Name(),
+		string(recommendedReplicas.Source),
+		le.JoinLeaderValue,
+	)
+
+	return &model.ScalingValues{
+		Horizontal: &recommendedReplicas,
+	}, nil
+}
+
+func (r resourceRecommenderSettings) recommend(currentTime time.Time, pods []*workloadmeta.KubernetesPod, queryResult loadstore.QueryResult, currentReplicas float64) (int32, utilizationResult, error) {
```

Comment:
I think these methods should not be attached to the `resourceRecommenderSettings`. The settings should be passed in as parameter to either a free function or to a `replicaCalculator` function